tests/util/caches/test_descriptors.py

```diff
@@ -84,3 +91,86 @@ def fn(self, arg1, arg2):
         r = yield obj.fn(2, 5)
         self.assertEqual(r, 'chips')
         obj.mock.assert_not_called()
+
+    def test_cache_logcontexts(self):
+        """Check that logcontexts are set and restored correctly when
+        using the cache."""
+        complete_lookup = [False]
+
+        class Cls(object):
+            @descriptors.cached()
+            def fn(self, arg1):
+                @defer.inlineCallbacks
+                def inner_fn():
+                    while not complete_lookup[0]:
+                        yield async.sleep(0.001)
```

Comment:
because that would run the second half of the lookup function in the logcontext provided for `.callback`. The point here was to replicate as closely and as simply as possible the standard yield-for-something-which-will-take-a-while pattern. 

Of course I could have done that with other things like PreserveLoggingContext, but then I would have had more moving parts which I had to trust.